### ISSUE TITLE
[HOLD for payment 2023-10-23] [$500] Android - Grant Camera Permission button not working on first click

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#27442]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Pre Requisite - Make sure camera permission is declined
1. Open FAB, Click Request money -> Scan Receipt
2. Click on Grant Permission.
3. If Permission Pop up open, deny permission.
4. Repeat 2 and 3 until redirected to settings, Close app and remove from recent
5. Open App ,  Open FAB, Click Request money -> Scan Receipt
6. Click on Grant Permission

## Expected Result:
Linked to App settings (as request permission is blocked by OS)

## Actual Result:
Clicking button does nothing on first click.

## Workaround:
Unknown 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.70.2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/bef52681-60b5-48c7-9433-6a97fc814d9d

https://github.com/Expensify/App/assets/93399543/c83cdeb0-93dd-4aa6-8c03-4972b9ffa35e

**Expensify/Expensify Issue URL:**
**Issue reported by:** @ishpaul777
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1694451532440139

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019c4a398f728939bc</li>
        <li>Upwork Job ID: 1702319435415248896</li>
        <li>Last Price Increase: 2023-09-28</li>
<li>Automatic offers: </li>
<ul>
<li>Ollyws | Reviewer | 26980712</li>
<li>zoontek | Contributor | 26980715</li>
<li>ishpaul777 | Reporter | 26980718</li>
<li>ntdiary | Contributor | 26981787</li>
</ul></li>
    </ul>
</details>

**Comment by akinwale** [comment]
This may be a duplicate of https://github.com/Expensify/App/issues/26499.

**Comment by ishpaul777** [comment]
I was assigned to the https://github.com/Expensify/App/issues/26499. I have to put the PR on Hold, as I was blocked by the same issue raised here, Another PR with similar changes I suggested https://github.com/Expensify/App/issues/26499 got merged recently. So this issue is not a dupe, but a regression from the PR  - https://github.com/Expensify/App/pull/26231

**Comment by trjExpensify** [comment]
But isn't #26499 the same problem as this issue, regardless of whether #26499 is a regression of PR - #26231?

**Comment by ishpaul777** [comment]
No the issue in https://github.com/Expensify/App/issues/26499 was that user always redirected to Settings page and a native modal never for asking permission never open, https://github.com/Expensify/App/issues/26499 was created to fix that. But Work was already in progress to solve this issue in PR https://github.com/Expensify/App/issues/26499, https://github.com/Expensify/App/issues/26499 was a Dupe. The issue here is when the user click on "Grant Permission" it does not work for first time if user has declined the permission request in past. Hope I am clear here?

**Comment by Julesssss** [comment]
>  So this issue is not a dupe, but a regression from the PR 

Not sure I would call it a regression, as that PR introduced the feature, just in a broken state.

But I do agree this is a separate issue. @trjExpensify it's slightly different: 
- #26499 -->  When asking for permission, we are not using the native Android Modal
- #27442 -->  We are not asking for permission at all

**Comment by Julesssss** [comment]
Though perhaps both will have the same solution. [Here's](https://github.com/Expensify/App/issues/26499#issuecomment-1706745327) the most recent plan of action

**Comment by trjExpensify** [comment]
Okay, so this would be an eligible bug report. Do you want to hold this issue on #26499 going upstream?

**Comment by Julesssss** [comment]
It's going to be a little bit confusing, but I'd actually say we can work on both in paralell. 

@ishpaul777 it would be great if you wanted to work on both issues together? Please make a proposal for this one if youre interested.

**Comment by ishpaul777** [comment]
Unfortunatley, I am unable to find a satifactory solution for both the issue yet. Will make a proposal once i find a solution.

**Comment by trjExpensify** [comment]
@ishpaul777 any luck?

**Comment by ishpaul777** [comment]
@Julesssss @trjExpensify 
I spend some time looking into it, I think the best we can do is persist the blocked state for camera permission in onyx so when the permission is blocked we will persist the state in Onyx. Again when press "grant permission" we will check if permission is blocked by OS.

### Changes -

- Add a action to set the Onyx key in action https://github.com/Expensify/App/blob/main/src/libs/actions/Receipt.ts
```jsx
function persistBlockedPermission(isCameraBlocked: boolean) {
    Onyx.merge(ONYXKEYS.PERMISSION_BLOCKED, {
        isCameraPermissionBlocked: isCameraBlocked,
    });
}
```
- When user click "grant permission" and deny permission multiple time until OS blocks, we will store the state when permission state is "Blocked" 

https://github.com/Expensify/App/blob/main/src/pages/iou/ReceiptSelector/index.native.js#L151
```JS
 const askForPermissions = () => {
        // There's no way we can check for the BLOCKED status without requesting the permission first
        // https://github.com/zoontek/react-native-permissions/blob/a836e114ce3a180b2b23916292c79841a267d828/README.md?plain=1#L670
        if (permissions === RESULTS.BLOCKED || isAndroidBlockedPermissionRef.current) {
            Linking.openSettings();
        } else if (permissions === RESULTS.DENIED) {
            CameraPermission.requestCameraPermission().then((permissionStatus) => {
                setPermissions(permissionStatus);
                if (permissionStatus !== RESULTS.BLOCKED) {
                    isAndroidBlockedPermissionRef.current = false;
                    Receipt.persistBlockedPermission(false);
                    return;
                }
                isAndroidBlockedPermissionRef.current = true;
                Receipt.persistBlockedPermission(true);
            });
        }
```
- if the app has come back from background, refresh the permissions status and remove set the blocked permission status to `false`.  https://github.com/Expensify/App/blob/main/src/pages/iou/ReceiptSelector/index.native.js#L100
```diff
useEffect(() => {
        const subscription = AppState.addEventListener('change', (nextAppState) => {
            if (appState.current.match(/inactive|background/) && nextAppState === 'active') {
                CameraPermission.getCameraPermissionStatus().then((permissionStatus) => {
                    setPermissions(permissionStatus);
 +                   if (permissionStatus === RESULTS.GRANTED) {
 +                       isAndroidBlockedPermissionRef.current = false;
 +                      Receipt.persistBlockedPermission(false);
 +                   }
                });
            }
```

- Set key to false when we check for camera permission on component mount and result is 'granted'.
https://github.com/Expensify/App/blob/main/src/pages/iou/ReceiptSelector/index.native.js#L224
```diff
 CameraPermission.getCameraPermissionStatus().then((permissionStatus) => {
        setPermissions(permissionStatus);
+        if (permissionStatus === RESULTS.GRANTED) {
+            isAndroidBlockedPermissionRef.current = false;
+           Receipt.persistBlockedPermission(false);
+        }
    });
```

**Comment by trjExpensify** [comment]
@Julesssss thoughts on the above?

**Comment by Julesssss** [comment]
> I think the best we can do is persist the blocked state

Thanks, it's a reasonable idea, but sadly it complicates the flow by adding two completely separate conditions 🙁 

I still think the [best path forward here](https://github.com/Expensify/App/issues/26499#issuecomment-1706745327) is to fix this in the upstream library or switch library. Currently it doesn't match the behaviour of native apps.

**Comment by trjExpensify** [comment]
Let's go upstream then! @ishpaul777 you want to do that? @Julesssss who owns this library to connect them?

**Comment by ishpaul777** [comment]
Sadly I don't have enough experience working with the react-native library, i'll have to unassign myself from #26499

**Comment by Julesssss** [comment]
No worries, thanks for helping to look into this.

**Comment by trjExpensify** [comment]
@Julesssss can you confirm what library this is?

**Comment by Julesssss** [comment]
Hey again @zoontek 👋 

Here's another issue you might be interested in, linked to your react-native-permissions library. There a second linked [issue](https://github.com/Expensify/App/issues/26499#issuecomment-1706191379), and here is what we'd like to achieve ideally: 

**Regular flow**
- iOS: Button pressed > native permission prompt displayed
- Android: Button pressed > native permission prompt displayed

**Permission blocked**
- iOS: Button pressed > take to app settings page
- Android: Button pressed > take to app settings page

**Comment by zoontek** [comment]
@Julesssss Sure, let me check tonight, but this should be easily doable. I will come back with a proposal for both.

**Comment by trjExpensify** [comment]
Nice one, @zoontek! ❤️

**Comment by zoontek** [comment]
@Julesssss @trjExpensify Here's the fix: https://github.com/zoontek/App/compare/main...fix-camera-permission-prompt

- At start, the permission status in `undefined` (to avoid rendering an invalid UI until the check is done)
- We run a `useEffect` to check the status and store it in our state. For camera permission, the result could be `unavailable`, `denied` or `granted` (not `blocked`, as it's a `check()`)
- If it's `granted`, display the camera. Done! ✨
- If it's anything else, display the request button.
- On request button press, systematically perform a `request()` call. Store the result in our state.
- If the `request` returned `blocked`, open the app settings.
- When the app gain focus (`AppState status === active`), refresh our state to handle permissions status change.

Possible bonus improvement: You currently don't handle the case where the status is `unavailable` (could occur if the device don't have a camera, which is very, very rare). You could display an error alert in such cases (after request button press), like this:

```js
if (status === RESULTS.BLOCKED) {
  Linking.openSettings();
}
if (status === RESULTS.UNAVAILABLE) {
  Alert.alert(…)
}
```

**Comment by zoontek** [comment]
Note that persisting the blocked status is not a good solution. If at some point, it's `blocked`, we store it, then the user go to the app settings to set `Ask next time`, the `check()` will return `denied`, and as `blocked` is stored, you will determine that you cannot request it…except you can.

More details, if you want to dig a bit: https://github.com/zoontek/react-native-permissions/pull/696

**Comment by Julesssss** [comment]
Sounds good to me, thanks! @Ollyws can you please give the proposal a review?

**Comment by zoontek** [comment]
Following @ntdiary comment on #26499 (which will probably closed), I think we can continue the discussion here. Like I said [here](https://github.com/Expensify/App/issues/26499#issuecomment-1740916689), the proposal fixes this issue, but we could go for the extra mile on Android by:

- Showing an alert to explain that the user will be redirected to settings and he needs to activate the camera permission
- Showing an alert a second request try ([it uses shouldShowRequestPermissionRationale, like explained in the android permission request flow](https://developer.android.com/training/permissions/requesting#explain))

This could be done later, after this fix merge, as it's more of an UX improvement.

**Comment by zoontek** [comment]
I updated the PR to show the alert when the permission is `blocked`, saying that the user that he can be redirected to the settings to enable camera permission. IMHO the UX is kinda great.

**Comment by zoontek** [comment]
@Julesssss I applied for both jobs and created the PR here: https://github.com/Expensify/App/pull/28619

**Comment by Julesssss** [comment]
Can be paid out today

**Comment by trjExpensify** [comment]
Okay, so keeping these jobs separate for following the audit trail later if needs be. Confirming the payment summary as follows:

- $50 to @ishpaul777 for the bug report (paid!)
- $500 to @zoontek for the bug fix (Mathieu, waiting for you to [accept the offer](https://www.upwork.com/nx/wm/offer/26980715)).
- $500 to @ntdiary for the C+ review (paid!)

No #urgency bonus applied as per [here](https://github.com/Expensify/App/issues/27442#issuecomment-1757111811).

**Comment by zoontek** [comment]
@trjExpensify Sorry, I accepted it.

**Comment by trjExpensify** [comment]
Perfect, settled up with everyone. Closing!
